File: ui/dev-server.py
#!/usr/bin/env python3
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from json import dumps
from mimetypes import guess_type
from os import environ
from pathlib import Path
from urllib.parse import unquote, urlparse

UI_DIR = Path(__file__).resolve().parent
REPO_DIR = UI_DIR.parent
ENV_PATHS = [REPO_DIR.parent / ".env", REPO_DIR / ".env"]


def normalize_env_key(key):
    return key.replace("-", "_").upper()


def unquote_env(value):
    trimmed = value.strip()
    if len(trimmed) >= 2 and trimmed[0] == trimmed[-1] and trimmed[0] in ("'", '"'):
        return trimmed[1:-1]
    return trimmed


def read_env():
    env = {}
    raw_values = []
    for env_path in ENV_PATHS:
        if not env_path.exists():
            continue
        for line in env_path.read_text(encoding="utf-8").splitlines():
            trimmed = line.strip()
            if not trimmed or trimmed.startswith("#"):
                continue
            if trimmed.startswith("export "):
                trimmed = trimmed[7:].strip()
            if "=" in trimmed:
                key, value = trimmed.split("=", 1)
                key = normalize_env_key(key.strip())
                value = unquote_env(value)
                if value or key not in env:
                    env[key] = value
            else:
                raw_values.append(unquote_env(trimmed))
    if not first_value(env, ["YOUTUBE_API_KEY", "YOUTUBE_DATA_API_KEY", "YT_API_KEY", "YT_PUBLIC_API_KEY"]):
        env["YOUTUBE_API_KEY"] = raw_values[0] if raw_values else ""
    return env


def first_value(source, keys):
    for key in keys:
        value = source.get(normalize_env_key(key), "")
        if value:
            return value
    return ""


def frontend_config():
    env = read_env()
    return {
        "youtubeApiKey": first_value(env, ["YOUTUBE_API_KEY", "YOUTUBE_DATA_API_KEY", "YT_API_KEY", "YT_PUBLIC_API_KEY"]),
        "resolverBase": first_value(env, ["RESOLVER_URL", "YTDLP_RESOLVER_URL", "API_BASE_URL"]) or "http://3.121.216.41",
        "resolverKey": first_value(env, ["RESOLVER_API_KEY", "YTDLP_API_KEY", "API_KEY", "GENERATED_API_KEY"]),
        "googleClientId": first_value(env, ["GOOGLE_CLIENT_ID", "GOOGLE_OAUTH_CLIENT_ID"]),
    }


class UiHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path == "/config.js":
            self.send_config()
            return

        route = "/home.html" if parsed.path == "/" else parsed.path
        relative = unquote(route).lstrip("/")
        file_path = (UI_DIR / relative).resolve()
        if not str(file_path).startswith(str(UI_DIR)):
            self.send_text(403, "Forbidden")
            return
        if not file_path.exists() or not file_path.is_file():
            self.send_text(404, "Not found")
            return

        data = file_path.read_bytes()
        content_type = guess_type(str(file_path))[0] or "application/octet-stream"
        if content_type.startswith("text/") or file_path.suffix in {".js", ".json"}:
            content_type = f"{content_type}; charset=utf-8"
        self.send_response(200)
        self.send_header("content-type", content_type)
        self.send_header("cache-control", "no-store")
        self.end_headers()
        self.wfile.write(data)

    def send_config(self):
        body = f"window.YT_FRONTEND_CONFIG = {dumps(frontend_config())};\n".encode("utf-8")
        self.send_response(200)
        self.send_header("content-type", "text/javascript; charset=utf-8")
        self.send_header("cache-control", "no-store")
        self.end_headers()
        self.wfile.write(body)

    def send_text(self, status, body):
        data = body.encode("utf-8")
        self.send_response(status)
        self.send_header("content-type", "text/plain; charset=utf-8")
        self.send_header("cache-control", "no-store")
        self.end_headers()
        self.wfile.write(data)

    def log_message(self, format, *args):
        print(f"{self.address_string()} - {format % args}")


def main():
    host = environ.get("HOST", "127.0.0.1")
    port = int(environ.get("PORT", "8080"))
    server = ThreadingHTTPServer((host, port), UiHandler)
    print(f"Frontend UI running at http://{host}:{port}/")
    server.serve_forever()


if __name__ == "__main__":
    main()

File: ui/dev-server.mjs
import { createServer } from "node:http";
import { readFile } from "node:fs/promises";
import { existsSync } from "node:fs";
import { extname, join, normalize, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const uiDir = fileURLToPath(new URL(".", import.meta.url));
const repoDir = resolve(uiDir, "..");
const envPaths = [resolve(repoDir, "..", ".env"), join(repoDir, ".env")];
const port = Number(process.env.PORT || 8080);
const host = process.env.HOST || "127.0.0.1";

const contentTypes = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".webp": "image/webp",
};

const server = createServer(async (request, response) => {
  try {
    const requestUrl = new URL(request.url || "/", `http://${request.headers.host || "localhost"}`);

    if (requestUrl.pathname === "/config.js") {
      await sendConfig(response);
      return;
    }

    const pathname = requestUrl.pathname === "/" ? "/home.html" : requestUrl.pathname;
    const filePath = resolve(uiDir, `.${pathname}`);
    if (!filePath.startsWith(uiDir)) {
      sendText(response, 403, "Forbidden");
      return;
    }

    if (!existsSync(filePath)) {
      sendText(response, 404, "Not found");
      return;
    }

    const data = await readFile(filePath);
    const extension = extname(normalize(filePath)).toLowerCase();
    response.writeHead(200, {
      "content-type": contentTypes[extension] || "application/octet-stream",
      "cache-control": "no-store",
    });
    response.end(data);
  } catch (error) {
    sendText(response, 500, error instanceof Error ? error.message : "Server error");
  }
});

server.listen(port, host, () => {
  console.log(`Frontend UI running at http://${host}:${port}/`);
});

async function sendConfig(response) {
  const env = await readEnv();
  const youtubeApiKey = firstValue(env, [
    "YOUTUBE_API_KEY",
    "YOUTUBE_DATA_API_KEY",
    "YT_API_KEY",
    "YT_PUBLIC_API_KEY",
  ]);
  const resolverBase = firstValue(env, ["RESOLVER_URL", "YTDLP_RESOLVER_URL", "API_BASE_URL"])
    || "http://3.121.216.41";
  const resolverKey = firstValue(env, ["RESOLVER_API_KEY", "YTDLP_API_KEY", "API_KEY", "GENERATED_API_KEY"]);
  const googleClientId = firstValue(env, ["GOOGLE_CLIENT_ID", "GOOGLE_OAUTH_CLIENT_ID"]);

  const config = {
    youtubeApiKey,
    resolverBase,
    resolverKey,
    googleClientId,
  };

  response.writeHead(200, {
    "content-type": "text/javascript; charset=utf-8",
    "cache-control": "no-store",
  });
  response.end(`window.YT_FRONTEND_CONFIG = ${JSON.stringify(config)};\n`);
}

async function readEnv() {
  const env = {};
  const rawValues = [];

  for (const envPath of envPaths) {
    if (!existsSync(envPath)) {
      continue;
    }

    const text = await readFile(envPath, "utf8");
    for (const line of text.split(/\r?\n/)) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#")) {
        continue;
      }

      const match = /^(?:export\s+)?([A-Za-z0-9_-]+)\s*=\s*(.*)$/.exec(trimmed);
      if (match) {
        const key = normalizeEnvKey(match[1]);
        const value = unquote(match[2]);
        if (value || !Object.hasOwn(env, key)) {
          env[key] = value;
        }
        continue;
      }

      rawValues.push(unquote(trimmed));
    }
  }

  if (!firstValue(env, ["YOUTUBE_API_KEY", "YOUTUBE_DATA_API_KEY", "YT_API_KEY", "YT_PUBLIC_API_KEY"])) {
    env.YOUTUBE_API_KEY = rawValues[0] || "";
  }

  return env;
}

function firstValue(source, keys) {
  for (const key of keys) {
    const normalizedKey = normalizeEnvKey(key);
    if (source[normalizedKey]) {
      return source[normalizedKey];
    }
  }
  return "";
}

function normalizeEnvKey(key) {
  return key.replace(/-/g, "_").toUpperCase();
}

function unquote(value) {
  const trimmed = value.trim();
  if (
    (trimmed.startsWith('"') && trimmed.endsWith('"'))
    || (trimmed.startsWith("'") && trimmed.endsWith("'"))
  ) {
    return trimmed.slice(1, -1);
  }
  return trimmed;
}

function sendText(response, statusCode, body) {
  response.writeHead(statusCode, {
    "content-type": "text/plain; charset=utf-8",
    "cache-control": "no-store",
  });
  response.end(body);
}
